Consistent schema across partitions

muhamadazmy opened the pull request:
Consistent schema across partitions

Summary:
This PR make sure that PPs have a consistent service schema by writing
and upsert-schema record to bifrost. The leader processor makes sure
to write this record to bifrost when schema changes are detected

muhamadazmy commented on crates/wal-protocol/src/lib.rs:
```diff
@@ -242,397 +249,3 @@ impl MatchKeyQuery for Envelope {
         self.record_keys().matches_key_query(query)
     }
 }
-
```
This is currently not user (and will never be) a new draft for envelope redesign can be checked [here](https://github.com/restatedev/restate/pull/3854)

muhamadazmy commented:
@slinkydeveloper of course. Not urgent. let's discuss how we going to maintain the schema type. moving forward

slinkydeveloper reviewed (commented):
LGTM, as discussed offline, we need to verify what happens with large schemas and how we want to mitigate eventual issues in that direction.

AhmedSoliman reviewed (changes requested):
As it stands, this cannot be merged unless writes are gated with a check for restate version >= 1.7

AhmedSoliman reviewed (commented):
We are very close :)

muhamadazmy committed:
Consistent schema across partitions

Summary:
This PR make sure that PPs have a consistent service schema by writing
and upsert-schema record to bifrost. The leader processor makes sure
to write this record to bifrost when schema changes are detected

AhmedSoliman commented:
it turns out that this doesn't persist the schema in the partition store. Perhaps it was a follow up that was never implemented?

AhmedSoliman commented:
Sending a fix in https://github.com/restatedev/restate/pull/5000
